tests/test_engine.py

```diff
@@ -76,6 +99,36 @@ def test_has_table(self):
         d = engine.has_table('mytable')
         assert self.successResultOf(d) is True
 
+    def test_pooled_connection_reuse(self):
+        table = Table('my_pool_table', sqlalchemy.MetaData(),
+                      Column('id', Integer, primary_key=True),
+                      Column('name', String, nullable=False))
+        data = [{'id': 0,
+                 'name': 'a'},
+                {'id': 1,
+                 'name': 'b'},
+                {'id': 2,
+                 'name': 'c'}]
+        # max_overflow kwarg broken, but defaults to 10
+        engine = create_engine()
+        engine.execute(CreateTable(table))
+        engine.execute(table.insert(), data)
+        pool_size = engine._engine.pool.size
+        if callable(pool_size):
+            pool_size = pool_size()
+        if isinstance(engine._engine.pool, sqlalchemy.pool.QueuePool):
+            pool_size += engine._engine.pool._max_overflow
+        arbitrary = Random()
+        for _ in range(pool_size + 1):
+            entry = arbitrary.choice(list(range(3)))
+            d = engine.execute(table.select().where(
+                table.c.id == data[entry]['id']))
+
+            result = self.successResultOf(d)
```

Comment:
Just a clarifying note for future us; connection lifecycle looks like:

1. Every engine.execute() retrieves a free connection from pool
2. If there are no rows in the result set, the connection is released immediately (this is why UPDATE was calling checkin)
3. If ResultProxy.close() is called, the connection is released.
4. Otherwise, whenever the GC finalizes the result (immediately on CPython, rarely in Pypy) the connection is released via weakref.ref callback.